File: php/application/models/audio_clip_model.php
<?php

class Audio_clip_model extends CI_Model {
    
    public function get_all( $offset = 0, $limit = 20 )
    {
        $audio_clips = $this->gramofonclient->get_audio_clips($offset, $limit);
        
        if ( !empty($audio_clips) ) {            
            foreach ( $audio_clips as &$clip ) {
                $clip->created = strtotime($clip->created);
            }
        }
        
        return $audio_clips;
    }
    
    public function get_user_audio_clips( $user_id, $offset = 0, $limit = 20 )
    {        
        $audio_clips = $this->gramofonclient->get_user_audio_clips($user_id, $offset, $limit);
        
        if ( !empty($audio_clips) ) { 
            if ( !is_array($audio_clips) ) {
                $audio_clips = array( $audio_clips );
            } 
            
            foreach ( $audio_clips as &$clip ) {
                $clip->created = strtotime($clip->created);
            }
        }
        
        return $audio_clips;
    }
    
    public function get_audio_clip( $id )
    {
        $clip = $this->gramofonclient->get_audio_clip($id);
        
        if ( !empty($clip) ) {             
            $clip->created = strtotime($clip->created);
        }
        
        return $clip;
    }
    
    
    public function update_audio_clip( $id, $data )
    {
        $clip = $this->gramofonclient->update_audio_clip($id, $data );
        
        return $clip;
    }
    
}

File: ios/gramofon-v1/gramofon-v1/Models/AudioClipModel.m
//
//  AudioClipModel.m
//  gramofon
//
//  Created by Dan Trenz on 4/6/13.
//  Copyright (c) 2013 gramofon. All rights reserved.
//

#import "AudioClipModel.h"
#import "HTTPRequest.h"
#import "AudioClip.h"
#import "User.h"

@implementation AudioClipModel

+ (void)getAudioClips:(int)offset itemCount:(int)limit complete:(void (^)(NSData *))completeCallback
{
    NSString *url = [NSString stringWithFormat:@"http://api.gramofon.co/clips?offset=%i&limit=%i", offset, limit];
    
    // asynchrous loading of clips w/ complete callback handler
    [[HTTPRequest sharedInstance] doAsynchRequest:@"GET"
                                       requestURL:url
                                    requestParams:nil
                                  completeHandler:^(NSURLResponse *response, NSData *data, NSError *error)
                                  {
                                      if ( ! error ) {
                                          if ( completeCallback != nil ) {
                                              completeCallback(data);
                                          }
                                      } else {
                                          NSLog(@"Error: %@", [error localizedDescription]);
                                      }
                                  }];
}

+ (void)getAudioClipsByUser:(NSNumber *)user_id itemOffset:(int)offset itemCount:(int)limit complete:(void (^)(NSData *))completeCallback
{
    NSString *url = [NSString stringWithFormat:@"http://api.gramofon.co/users/%@/clips?offset=%i&limit=%i", user_id, offset, limit];
    
    // asynchrous loading of clips w/ complete callback handler
    [[HTTPRequest sharedInstance] doAsynchRequest:@"GET"
                                       requestURL:url
                                    requestParams:nil
                                  completeHandler:^(NSURLResponse *response, NSData *data, NSError *error)
                                  {
                                      if ( ! error ) {
                                          if ( completeCallback != nil ) {
                                              completeCallback( data );
                                          }
                                      } else {
                                          NSLog(@"Error: %@", [error localizedDescription]);
                                      }
                                  }];
}

+ (void)uploadAudioClip;
{
//    NSDictionary *params = [NSDictionary dictionaryWithObjectsAndKeys:
//                               [AudioClip sharedInstance].title, @"clip[title]",
//                               [AudioClip sharedInstance].longitude, @"clip[longitude]",
//                               [AudioClip sharedInstance].latitude, @"clip[latitude]",
//                               [AudioClip sharedInstance].venue, @"clip[venue]",
//                               [User sharedInstance].user_id, @"clip[user_id]", nil];
    NSMutableDictionary *params = [NSMutableDictionary dictionary];
    
    // required params
    params[@"clip[user_id]"] = [User sharedInstance].user_id;
    params[@"clip[title]"]   = [AudioClip sharedInstance].title;
    
    // optional location params
    if ( [AudioClip sharedInstance].longitude && [AudioClip sharedInstance].latitude ) {
        params[@"clip[longitude]"] = [AudioClip sharedInstance].longitude;
        params[@"clip[latitude]"]  = [AudioClip sharedInstance].latitude;
    }
    
    // optional venue params
    if ( [AudioClip sharedInstance].venue ) {
        params[@"clip[venue]"]  = [AudioClip sharedInstance].venue;        
    }
    NSLog(@"%@", params);
    NSDictionary *params2 = [NSDictionary dictionaryWithDictionary:params];
    
    [[HTTPRequest sharedInstance] uploadFile:@"http://api.gramofon.co/clips"
                                    fileName:[AudioClip sharedInstance].fileName
                                    fileData:[AudioClip sharedInstance].fileData
                                  postParams:params2];
}

+ (void)deleteAudioClip:(NSNumber *)clip_id complete:(void (^)(void))completeCallback
{
    NSString *url = [NSString stringWithFormat:@"http://api.gramofon.co/clips/%@", clip_id];
    
    // asynchrous loading of clips w/ complete callback handler
    [[HTTPRequest sharedInstance] doAsynchRequest:@"DELETE"
                                       requestURL:url
                                    requestParams:nil
                                  completeHandler:^(NSURLResponse *response, NSData *data, NSError *error)
                                  {
                                      if ( ! error ) {
                                          if ( completeCallback != nil ) {
                                              completeCallback();
                                          }
                                      } else {
                                          NSLog(@"Error: %@", [error localizedDescription]);
                                      }
                                  }];
}

@end
